### PR TITLE
phi node: filter multiple entries for same predecessor

### LINKED ISSUE #3865: phi node with multiple entries for same predecessor not handled correctly
This is kind of a weird edge case, but in LLVM IR, you can have multiple entries for the same predecessor block in a phi node if a switch or branch has duplicate destinations. This was reduced from a real failure in the OpenCL.jl test suite when running with the translator backend.

```llvm
target datalayout = "e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-G1"
target triple = "spir64-unknown-unknown"

define spir_kernel void @f() {
entry:
  br i1 undef, label %L815, label %L815

L815:                                             ; preds = %entry, %entry
  %v = phi i64 [ 0, %entry ], [ 0, %entry ]
  ret void
}
```

SPIRV-LLVM-Translator translates this verbatim:

```spirv
; SPIR-V
; Version: 1.0
; Generator: Khronos LLVM/SPIR-V Translator; 14
; Bound: 12
; Schema: 0
               OpCapability Addresses
               OpCapability Kernel
               OpCapability Int64
          %1 = OpExtInstImport "OpenCL.std"
               OpMemoryModel Physical64 OpenCL
               OpEntryPoint Kernel %4 "f"
               OpSource Unknown 0
               OpName %entry "entry"
               OpName %L815 "L815"
               OpName %v "v"
      %ulong = OpTypeInt 64 0
    %ulong_0 = OpConstant %ulong 0
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
       %bool = OpTypeBool
          %8 = OpUndef %bool
          %4 = OpFunction %void None %3
      %entry = OpLabel
               OpBranchConditional %8 %L815 %L815
       %L815 = OpLabel
          %v = OpPhi %ulong %ulong_0 %entry %ulong_0 %entry
               OpReturn
               OpFunctionEnd

```

which then fails validation with:
```
> spirv-val test.spv
error: line 21: OpPhi's number of incoming blocks (2) does not match block's predecessor count (2).
  %v = OpPhi %ulong %ulong_0 %entry %ulong_0 %entry
```

I am not 100% sure whether this is just a bug in spirv-val, but the LLVM backend at least produces `%v = OpPhi %ulong %ulong_0 %entry` in line 21 here (I actually had to make `%v` the return and pass `-O0` for `llc` to not just optimize away the phi node)

### PR DESCRIPTION
**Pull request by simeonschaub**
fixes #3865

### CODE REVIEW

File: test/phi-duplicate-predecessors.ll [hunk 1]
```diff
@@ -0,0 +1,22 @@
+; REQUIRES: spirv-dis
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv %t.bc -o %t.spv
+; RUN: spirv-val %t.spv
+; RUN: spirv-dis %t.spv | FileCheck %s
+
+; CHECK: %entry = OpLabel
+; CHECK: OpBranchConditional %[[NULL:[0-9]+]] %L815 %L815
+; CHECK: %L815 = OpLabel
+; CHECK: %v = OpPhi %ulong %ulong_0 %entry
```
Comment (MrSidims) [inline test/phi-duplicate-predecessors.ll]:
```suggestion
; CHECK: %[[Entry:[a-z0-9_]+]] = OpLabel
; CHECK: OpBranchConditional %[[#NULL:]] %[[Lable:[a-z0-9_]+]] %[[Lable]]
; CHECK: %[[Lable]] = OpLabel
; CHECK: %{{.*}} = OpPhi %{{.*}} %{{.*}} %[[Entry]]
```

File: test/phi-duplicate-predecessors.ll [hunk 2]
```diff
@@ -0,0 +1,22 @@
+; REQUIRES: spirv-dis
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv %t.bc -o %t.spv
+; RUN: spirv-val %t.spv
+; RUN: spirv-dis %t.spv | FileCheck %s
```
Comment (MrSidims) [inline test/phi-duplicate-predecessors.ll]: Please also add llvm-spirv -r RUN and check, so we make sure, we don't have a bug on the reverse path.

**Review by MrSidims — APPROVED** [review]
LGTM, thank you!

File: test/phi-duplicate-predecessors.ll (hunk 1 above)
Comment (MrSidims) [reply test/phi-duplicate-predecessors.ll]: (tho I never remember string variable declaration in LIT spelling, better to double check)

**Comment by simeonschaub** [comment]
The In-tree failure looks real, I wonder if this assert might be too strict?

https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/880cbb08c8d3c260800feb53a1c18537e9ecdbdb/lib/SPIRV/libSPIRV/SPIRVInstruction.h#L1017

**Comment by simeonschaub** [comment]
Bump, any ideas what to do about the failing assertion? Should it just be deleted?

**Comment by MrSidims** [comment]
> Bump, any ideas what to do about the failing assertion? Should it just be deleted?

Apologies, will try to take a look by EOW.

**Comment by MrSidims** [comment]
@simeonschaub assert is legit, it reflects "#645: The True Label and False Label of an [OpBranchConditional](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpBranchConditional) must not be the same, starting with version 1.6.", see https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_changes_from_version_1_5 - not sure why spirv-val is passing, probably because generated module is 1.0 (but I'm always forgetting the exact mechanism the translator picks the version when none are specified ;-P) - would be nice to add to the test max version enforcement to 1.6.

Regarding what to do with it. IMHO - just drop? Such a branch is equivalent to an unconditional one - the condition is a plain value, evaluating it has no side effects - so emit OpBranch instead. WDYT?

**Comment by simeonschaub** [comment]
Makes sense! I have added special handling for conditional branches where both successors are the same and adjusted the test

**Comment by MrSidims** [comment]
Cool! I'm ramping down my contributions to llvm-spirv tool, so would be nice to get feedback from @svenvh and @vmaksimo

File: lib/SPIRV/SPIRVWriter.cpp [hunk 3]
```diff
@@ -2593,6 +2593,12 @@ LLVMToSPIRVBase::transValueWithoutDecoration(Value *V, SPIRVBasicBlock *BB,
                                LoopControl, Parameters, SuccessorTrue);
       }
     }
+    // Starting with SPIR-V 1.6, the True Label and False Label of an
+    // "OpBranchConditional" must not be the same, so emit an unconditional
+    // branch.
+    if (SuccessorTrue == SuccessorFalse)
+      return mapValue(V, BM->addBranchInst(SuccessorTrue, BB));
+
```
Comment (MrSidims) [inline lib/SPIRV/SPIRVWriter.cpp]: May be the assertion should also be guarded with IF SPIR_V_1_6 check, 

**UPD**: though, with this replacement it should never fire (which is exactly the purpose of assertions). So I don't have a strong opinion about it actually. May leave the assertion as is - your call :)

Comment (simeonschaub) [reply lib/SPIRV/SPIRVWriter.cpp]: My thought is that this was probably just left under-specified before SPIR-V 1.6, so it's probably good to assert nonetheless to prevent any issues down the road?

File: test/phi-duplicate-predecessors.ll [hunk 4]
```diff
@@ -0,0 +1,31 @@
+; REQUIRES: spirv-dis
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv %t.bc -o %t.spv
```
Comment (vmaksimo) [inline test/phi-duplicate-predecessors.ll]:
```suggestion
; RUN: llvm-spirv %s -o %t.spv
```
just a small suggestion

File: lib/SPIRV/SPIRVWriter.cpp [hunk 5]
```diff
@@ -2593,6 +2593,12 @@ LLVMToSPIRVBase::transValueWithoutDecoration(Value *V, SPIRVBasicBlock *BB,
                                LoopControl, Parameters, SuccessorTrue);
       }
     }
+    // Starting with SPIR-V 1.6, the True Label and False Label of an
+    // "OpBranchConditional" must not be the same, so emit an unconditional
+    // branch.
```
Comment (vmaksimo) [inline lib/SPIRV/SPIRVWriter.cpp]: The transform here is applied with any SPIR-V version, maybe we can reword the comment to avoid the confusion, e.g. like:
```
... emit an unconditional branch - always valid, required for 1.6+
```

**Review by vmaksimo — COMMENTED** [review]
LGTM, just few things to improve

File: lib/SPIRV/SPIRVWriter.cpp (hunk 5 above)
Comment (simeonschaub) [reply lib/SPIRV/SPIRVWriter.cpp]:
```suggestion
    // Starting with SPIR-V 1.6, the True Label and False Label of an
    // "OpBranchConditional" must not be the same, so emit an unconditional
    // branch - always valid, required for 1.6+
```

**Review by vmaksimo — APPROVED** [review]
